types/datetime.go

```diff
@@ -0,0 +1,292 @@
+	if c == '.' {
+		i++
+
+		if i+3 > length {
+			return Datetime{}, fmt.Errorf("%w: invalid millisecond", ErrDatetime)
+		}
+
+		if !(unicode.IsDigit(rune(s[i])) &&
+			unicode.IsDigit(rune(s[i+1])) &&
+			unicode.IsDigit(rune(s[i+2]))) {
+			return Datetime{}, fmt.Errorf("%w: invalid millisecond", ErrDatetime)
+		}
+
+		milli = 100*int(rune(s[i])-'0') + 10*int(rune(s[i+1])-'0') + int(rune(s[i+2])-'0')
+		i = i + 3
+	}
+
+	if i >= length {
+		return Datetime{}, fmt.Errorf("%w: expected timezone indicator", ErrDatetime)
+	}
+
+	c = rune(s[i])
+	if c == 'Z' { // "YYYY-MM-DDThh:mm:ssZ"
+		if i+1 == length {
+			t := time.Date(year, time.Month(month),
+				day, hour, minute, second,
+				int(time.Duration(milli)*time.Millisecond), time.UTC)
+			return Datetime{Value: t.UnixMilli()}, nil
+		}
+		return Datetime{}, fmt.Errorf("%w: unexpected trailer after timezone indicator", ErrDatetime)
+	}
+
+	// It must have an offset. +/-hhmm
+
+	if i+5 > length {
+		return Datetime{}, fmt.Errorf("%w: expected time offset", ErrDatetime)
+	} else if i+5 < length {
+		return Datetime{}, fmt.Errorf("%w: unexpected trailer", ErrDatetime)
+	}
+
+	sign := time.Duration(1)
+	c = rune(s[i])
+	if c == '-' {
+		sign = -sign
+	} else if c != '+' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(c))
+	}
+
+	i++
+
+	if !(unicode.IsDigit(rune(s[i])) &&
+		unicode.IsDigit(rune(s[i+1])) &&
+		unicode.IsDigit(rune(s[i+2])) &&
+		unicode.IsDigit(rune(s[i+3]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid time offset", ErrDatetime)
+	}
+
+	offsetH := time.Duration(10*int64(rune(s[i])-'0')+int64(rune(s[i+1])-'0')) * time.Hour
+	offsetM := time.Duration(10*int64(rune(s[i+2])-'0')+int64(rune(s[i+3])-'0')) * time.Minute
+
+	t := time.Date(year, time.Month(month), day,
+		hour, minute, second,
+		int(time.Duration(milli)*time.Millisecond), time.UTC)
+	t = t.Add(sign * (offsetH + offsetM))
+	return Datetime{Value: t.UnixMilli()}, nil
+}
+
+func (a Datetime) Equal(bi Value) bool {
+	b, ok := bi.(Datetime)
+	return ok && a == b
+}
+
+func (a Datetime) Less(bi Value) bool {
+	b, ok := bi.(Datetime)
+	return ok && a.Value < b.Value
+}
+
+func (a Datetime) LessEqual(bi Value) bool {
+	b, ok := bi.(Datetime)
+	return ok && a.Value <= b.Value
+}
+
+func (a Datetime) MarshalCedar() []byte {
+	return []byte(`datetime("` + a.String() + `")`)
+}
+
+func (a Datetime) String() string {
+	return time.UnixMilli(a.Value).UTC().Format("2006-01-02T15:04:05.000Z")
+}
+
+func (a *Datetime) UnmarshalJSON(b []byte) error {
+	var arg string
+	if len(b) > 0 && b[0] == '"' {
+		if err := json.Unmarshal(b, &arg); err != nil {
+			return errors.Join(errJSONDecode, err)
+		}
+	} else {
+		// NOTE: cedar supports two other forms, for now we're only supporting the smallest implicit and explicit form.
+		// The following are not supported:
+		// "datetime(\"1970-01-01T00:00:00Z\")"
+		// {"fn":"datetime","arg":"1970-01-01T00:00:00Z"}
+		var res extValueJSON
```

Comment:
As long as this is in parity with what we did for Decimal/IPAddr it's okay for now.  But maybe add an issue somewhere about the deficit.

Reply:
I'll just add it for duration/datetime, as they aren't hard to add.